File: day9/day9.rb
require "set"

Position = Struct.new(:i, :j) do
  def to(direction)
    di, dj = direction
    Position.new(i + di, j + dj)
  end

  def follow(other_position)
    di = other_position.i - i
    dj = other_position.j - j
    to([di.clamp(-1, 1), dj.clamp(-1, 1)])
  end
end

Rope = Struct.new(:positions) do
  def move(direction)
    positions[0] = positions[0].to(direction)
    (1...positions.size).each do |i|
      break unless is_too_far?(i)

      positions[i] = positions[i].follow(positions[i - 1])
    end
  end

  def tail
    positions.last
  end

  private

  def is_too_far?(i)
    p1, p2 = positions.slice(i - 1, 2)
    (p1.i - p2.i).abs > 1 || (p1.j - p2.j).abs > 1
  end
end

DIRECTIONS = {
  "R" => [ 1,  0],
  "L" => [-1,  0],
  "U" => [ 0, -1],
  "D" => [ 0,  1],
}

N = 10

zero = Position.new(0, 0)
rope = Rope.new([zero] * N)
tail_visited_positions = Set.new([rope.tail])

ARGF
  .each_line
  .map(&:chomp)
  .map(&:split)
  .each do |(d, n)|
    direction = DIRECTIONS.fetch(d)
    n.to_i.times do
      rope.move(direction)
      tail_visited_positions << rope.tail
    end
  end

puts tail_visited_positions.size

File: day9/day9.py
import sys

from dataclasses import dataclass
from typing import Iterator, Tuple

def clamp(i: int) -> int:
    return max(min(i, 1), -1)

Direction = Tuple[int, int]
Instruction = Tuple[Direction, int]

@dataclass(frozen=True)
class Position:
    i: int
    j: int

    def move_to(self, direction: Direction) -> "Position":
        di, dj = direction
        return Position(self.i + di, self.j + dj)

    def follow_other(self, other: "Position") -> "Position":
        di = other.i - self.i
        dj = other.j - self.j
        if abs(di) > 1 or abs(dj) > 1:
            return self.move_to(direction=(clamp(di), clamp(dj)))
        else:
            return self

def read_instructions() -> Iterator[Instruction]:
    directions: dict[str, Direction] = {
        "R": ( 1,  0),
        "L": (-1,  0),
        "U": ( 0, -1),
        "D": ( 0,  1),
    }
    with open(sys.argv[1], 'r') as file:
        for line in file:
            d, n = line.split(" ")
            yield (directions[d], int(n))

N = 10

zero = Position(0, 0)
positions = [zero] * N
tail_visited_positions = {zero}

for direction, number_of_moves in read_instructions():
    for _ in range(number_of_moves):
        positions[0] = positions[0].move_to(direction)
        for i in range(1, N):
            positions[i] = positions[i].follow_other(positions[i - 1])

        tail_visited_positions.add(positions[-1])

print(len(tail_visited_positions))
